alacritty_terminal/src/renderer/mod.rs

```diff
@@ -747,15 +748,25 @@ impl QuadRenderer {
             gl::EnableVertexAttribArray(0);
         }
 
-        // Draw visual bell
-        let rect = RenderRect::new(0., 0., props.width, props.height, visual_bell_color);
-        self.render_rect(&rect, visual_bell_intensity as f32, props);
-
         // Draw underlines and strikeouts
         for cell_line_rect in cell_line_rects {
             self.render_rect(&cell_line_rect, 255., props);
         }
 
+        // Change blending function for visual bell to blend only colors and keep alpha
+        unsafe {
+            gl::BlendFuncSeparate(gl::SRC_ALPHA, gl::ONE_MINUS_SRC_ALPHA, gl::SRC_ALPHA, gl::ONE);
```

Comment:
Underlines are a part of a text imho, so we don't want to make them have not 100% opacity.

Message bar is above background, so it's expected to me that it should be fully opaque.

I don't think that we want to blend colors and opacity separately in general. This is just a hack to blend colors with by opacity, preserving source alpha and setting 100% opacity to a bell.

Reply:
However it seems like we can just change our blending function in draw_rects to just this one.

Reply:
> Underlines are a part of a text imho, so we don't want to make them have not 100% opacity.

Sure, but that's mostly because we should set an alpha of 1.0 for single pixel underlines. Imagine having fuzzier and bigger underlines that are round at each side. That would likely produce cells that aren't 100% opaque and you wouldn't want to render those in a way that makes the background transparent.

This is mostly a theoretical problem since our underlines are never going to be that complex, but to correctly express our intent, the underlines should be rendered *above* the background, not merged with it. But same thing goes for a message bar if we would want to render it at 50% opacity, it should be composited above the background.

So I definitely do think that we want to blend colors and opacity separately for all currently used rects. It's just that with underlines and message bar we currently choose to have 100% opacity.

Reply:
I've already done this though, if you're about render_rects